tests/pytorch/test_fused_rope.py

```diff
@@ -41,20 +42,118 @@ def apply_rotary_pos_emb_thd(
         cu_seqlens(Tensor):  Cumulative sum of sequence lengths in a batch for `t`,
         with shape [b + 1] and dtype torch.int32.
         freqs (Tensor): Rotary Positional embedding tensor freq is of shape [max_s, 1, 1, d]
+        start_positions (Tensor): Tensor of shape [b] determining the beginning offsets
+                         of frequeuncies applied to  sequences.
 
     Returns:
         Tensor: Shape [t, h, d]. The input tensor after applying RoPE.
     """
     cu_seqlens = cu_seqlens // cp_size
     seqlens = (cu_seqlens[1:] - cu_seqlens[:-1]).tolist()
-    return torch.cat(
-        [
-            apply_rotary_pos_emb(
-                x.unsqueeze(1), _get_thd_freqs_on_this_cp_rank(cp_rank, cp_size, x, freqs)
-            )
-            for x in torch.split(t, seqlens)
-        ]
-    ).squeeze(1)
+    if cp_size > 1:
+        return torch.cat(
+            [
+                apply_rotary_pos_emb(
+                    x.unsqueeze(1), _get_thd_freqs_on_this_cp_rank(cp_rank, cp_size, x, freqs)
+                )
+                for x in torch.split(t, seqlens)
+            ]
+        ).squeeze(1)
+    else:
+        if start_positions is None:
+            return torch.cat(
+                [
+                    apply_rotary_pos_emb(x.unsqueeze(1), freqs[: x.size(0)])
+                    for x in torch.split(t, seqlens)
+                ]
+            ).squeeze(1)
+        else:
+            return torch.cat(
+                [
+                    apply_rotary_pos_emb(
+                        x.unsqueeze(1), freqs[start_positions[i] : (x.size(0) + start_positions[i])]
+                    )
+                    for i, x in enumerate(torch.split(t, seqlens))
+                ]
+            ).squeeze(1)
+
+
+def apply_rotary_pos_emb_with_start_positions(
```

Comment:
It's better to merged this with `apply_rotary_pos_emb` in `rope.py`.

Reply:
Done